File: 44-wildcard-matching/44-wildcard-matching.c

int n,m;
int d[2001][2001];
char *s, *p;
bool f(int i, int j) {
    int *r = &d[i][j];
    if (*r != -1) return *r;
    *r = 0;
    if (j == m) return *r = i == n;
    if (i == n) return *r = p[j] == '*' && f(i, j + 1);
    if (p[j] == '*') *r = f(i + 1, j) | f(i, j + 1);
    if (s[i] == p[j] || p[j] == '?') *r = f(i + 1, j + 1);
    return *r;
}
bool isMatch(char * ss, char * pp){
    s = ss;
    p = pp;
    n = strlen(ss);
    m = strlen(pp);
    memset(d, -1, 2001 * 2001 * sizeof(int));
    return f(0, 0);
}

File: 44-wildcard-matching/44-wildcard-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        n,m = len(s),len(p)
        dp = {(n, m): 1}
        def f(i, j):
            if (i,j) in dp:
                return dp[i,j]
            if j == m:
                dp[i, j] = i == n
                return dp[i, j]
            if i >= n:
                dp[i, j] = p[j] == '*' and f(i, j + 1)
                return dp[i, j]
            dp[i, j] = 0
            if p[j] == '*':
                dp[i,j] = f(i, j + 1) or f(i + 1, j)
            if p[j] in (s[i], '?'):
                dp[i, j] = f(i + 1, j + 1)
            return dp[i, j]

        #f(0, 0)
        #print(dp)
        return f(0, 0)